src/pressure/VCycleMG.cpp

```diff
@@ -753,154 +612,244 @@ void VCycleMG::Solve(Field &out, Field &tmp, Field const &b, size_t level, bool
+
+#pragma acc wait
+            res = sqrt(sum);
+            it++;
+        }  // end while
     }  // end data region
 
-// Diffusion step
-    std::string diffusionType = params->get("solver/pressure/diffusion/type");
-    if (diffusionType == DiffusionMethods::Jacobi) {
+}
+void VCycleMG::call_smooth_jacobi(Field &out, Field &tmp, Field const &b, const size_t level, bool sync) {
+    auto domain = Domain::getInstance();
+
+    // local variables and parameters for GPU
+    const real dx = domain->get_dx(level);
+    const real dy = domain->get_dy(level);
+    const real dz = domain->get_dz(level);
+
+    real *data_out = out.data;
+    real *data_tmp = tmp.data;
+
+    FieldType type = out.get_type();
+
+    BoundaryController *boundary = BoundaryController::getInstance();
+    boundary->apply_boundary(data_out, level, type, sync);
+
+    const real rdx2 = 1. / (dx * dx);
+    const real rdy2 = 1. / (dy * dy);
+    const real rdz2 = 1. / (dz * dz);
+
+    // preparation for diffusion step (alpha, beta)
+    const real alphaX = rdx2;
+    const real alphaY = rdy2;
+    const real alphaZ = rdz2;
+
+    const real rbeta = 2. * (alphaX + alphaY + alphaZ);
+    const real beta = 1. / rbeta;
+
+    tmp.copy_data(out);
 #pragma acc data present(out, tmp, b)
-        {
-            size_t it = 0;
-            const size_t max_it = static_cast<size_t> (params->get_int("solver/pressure/diffusion/max_solve"));
-            const real tol_res = params->get_real("solver/pressure/diffusion/tol_res");
-            real sum;
-            real res = 10000.;
-
-            while (res > tol_res && it < max_it) {
-                JacobiDiffuse::JacobiStep(level, out, tmp, b, alphaX, alphaY, alphaZ, beta, m_dsign, m_w, sync);
-                boundary->apply_boundary(out.data, level, type, sync);
-
-                sum = 0.;
-
-#pragma acc parallel loop independent present(out, tmp, d_iList[:bsize_i]) async
-                for (size_t j = start_i; j < end_i; ++j) {
-                    const size_t i = d_iList[j];
-                    res = b[i] - (rdx2 * (out[i - 1] - 2 * out[i] + out[i + 1])\
- + rdy2 * (out[i - Nx] - 2 * out[i] + out[i + Nx])\
- + rdz2 * (out[i - Nx * Ny] - 2 * out[i] + out[i + Nx * Ny])); //res = rbeta*(out[i] - tmp[i]);
-                    sum += res * res;
-                }
-                // info: in nvvp profile 8byte size copy from to device to/from pageable due to sum!
+    {
+        size_t it = 0;
+        for (int i = 0; i < m_n_relax; i++){
+            JacobiDiffuse::JacobiStep(level, out, tmp, b, alphaX, alphaY, alphaZ, beta, m_dsign, m_w, sync);
+            boundary->apply_boundary(data_out, level, type, sync);
+            Field::swap(tmp, out);
+            std::swap(data_tmp, data_out);
+            it++;
+        }  // end while
+
+        if (it % 2 != 0) {  // get data from tmp field when number of iterations is odd
+            out.copy_data(tmp);
+        }
+    } //end data region
```

Comment:
two spaces between code and comment are needed.

Reply:
b167247